### PR TITLE
fix: refresh page client state lost error.

### LINKED ISSUE #3845: Reloading via ctrl-r on the Settings page of a bot with an ejected runtime throws an error
Screen Shot 2020-08-13 at 6 47 41 PM

### PR DESCRIPTION
**Pull request by zhixzhan**
## Description

Current if re-fresh page in home/setting page, all client state are lost. Notice `navigation` are disabled:
Untitled5

other page (design/lg/lu) which url include `:projectId` don't have this issue, cause when refresh page, `projectId` is still in url.

### Changes made to fix this
1. Cache projectId at client session storage.
2. Auto-fetch project if current no project is opened.

## Task Item
close #3845 

## Screenshots

Refresh in home/setting page wouldn't lost states

Untitled4

### CODE REVIEW

File: Composer/packages/client/src/pages/setting/SettingsPage.tsx [hunk 1]
```diff
@@ -50,6 +52,14 @@ const SettingPage: React.FC<RouteComponentProps<{ '*': string }>> = () => {
...
   const { navigate } = useLocation();
 
+  // when fresh page, projectId in store are empty, no project are opened at client
+  // use cached projectId do fetch.
+  const cachedProjectId = useProjectIdCache();
+  useEffect(() => {
+    if (!projectId && cachedProjectId) {
+      fetchProjectById(cachedProjectId);
+    }
+  }, [projectId]);
```
Comment (yeze322) [inline Composer/packages/client/src/pages/setting/SettingsPage.tsx:62]: since it's duplicated, we can move it inside the shared hook.

Comment (zhixzhan) [reply Composer/packages/client/src/pages/setting/SettingsPage.tsx:62]: I think it's not been used in a lot place or complex enough. maybe leave it here would be more flexible.

Comment (yeze322) [reply Composer/packages/client/src/pages/setting/SettingsPage.tsx:62]: If the fix only applies to the page refresh scenario, the deps of `useEffect` can become `[]`?

Comment (zhixzhan) [reply Composer/packages/client/src/pages/setting/SettingsPage.tsx]: I moved fetch project logic to root router, better than do this in components.

File: Composer/packages/client/src/pages/home/Home.tsx [hunk 2]
```diff
@@ -136,10 +132,6 @@ const Home: React.FC<RouteComponentProps> = () => {
     },
   ];
 
-  useEffect(() => {
-    fetchRecentProjects();
```
Comment (zhixzhan) [inline Composer/packages/client/src/pages/home/Home.tsx:140]: move fetchRecentProjects to creationFlow to centralize resource fetch in sequence.

File: Composer/packages/client/src/pages/setting/dialog-settings/DialogSettings.tsx [hunk 3]
```diff
@@ -37,9 +36,6 @@ export const DialogSettings: React.FC<RouteComponentProps> = () => {
   const { setSettings, setLocale, addLanguageDialogBegin } = useRecoilValue(dispatcherState);
 
   const { languages, defaultLanguage } = settings;
-  useEffect(() => {
-    if (!botName) navigateTo('/home');
```
Comment (zhixzhan) [inline Composer/packages/client/src/pages/setting/dialog-settings/DialogSettings.tsx:41]: No bot opened , navigate to Home. No needed anymore cause we also fetchProject in setting page.

**Comment by boydc2014** [comment]
verified on my side

**Pull request merged**